experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts

```diff
@@ -0,0 +1,182 @@
+						// If it has children, return the special symbol indicating a list.
+						return node.length === 0 ? [] : [EmptyKey];
+					} else if (node instanceof Map) {
+						const unfilteredKeys = Array.from(node.keys()) as FieldKey[];
+						// Map proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the entry.
+						const filteredKeys = unfilteredKeys.filter(
+							(key) => node.get(key) !== undefined,
+						);
+						return filteredKeys;
+					} else {
+						// Assume record-like object
+						const objectedNode = node as Record<FieldKey, unknown>;
+
+						const unfilteredKeys = Object.keys(node) as FieldKey[];
+						// Object proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the key/value pair.
+						const filteredKeys = unfilteredKeys.filter((key) => {
+							const value = objectedNode[key];
+							return value !== undefined;
+						});
+						return filteredKeys;
+					}
+				default:
+					return [];
+			}
+		},
+		getFieldFromNode: (node: TNode, key: FieldKey): readonly TNode[] => {
+			if (node === undefined) {
+				throw new Error(
+					"Cursor adapter was passed an undefined node. This is unsupported.",
+				);
+			}
+
+			// Object.prototype.hasOwnProperty can return true for strings (ex: with key "0"), so we have to filter them out.
+			// Rather than just special casing strings, we can handle them with an early return for all primitives.
+			if (typeof node !== "object") {
+				return [];
+			}
+
+			if (node === null) {
+				return [];
+			}
+
+			if (Array.isArray(node)) {
+				// TODO: verify this
+				return key === EmptyKey || key === rootFieldKey ? node : [];
+			}
+
+			if (node instanceof Map) {
+				const value = (node as Map<FieldKey, TNode>).get(key);
+				return value === undefined ? [] : [value];
+			}
+
+			if (Object.prototype.hasOwnProperty.call(node, key)) {
+				const field = (node as Record<FieldKey, TNode>)[key];
+				return field === undefined ? [] : [field];
+			}
+
+			return [];
+		},
+	};
+}
+
+/**
+ * Construct a {@link CursorWithNode} from a {@link ProxyNode}.
+ *
+ * @param node - The tree being inserted, which will be mapped to the editable tree domain via {@link createProxyTreeAdapter}.
+ * @param context - The {@link TreeEntity.context} of the parent node under which the tree is being inserted.
+ * @param typeSet - The types allowed under the parent node under which the tree is being inserted.
+ * @param mode - The mode of cursor creation.
+ * Determines whether the cursor is initialized in "node" mode or "field" mode.
+ */
+export function cursorFromProxyTree<TNode extends ProxyNode<TreeNodeSchema, "javaScript">>(
+	node: TNode,
+	context: TreeDataContext,
+	typeSet: TreeTypeSet,
+	mode: "node" | "field" = "node",
```

Comment:
Is it appropriate to use `CursorLocationType` here, or is that a weird import?

Reply:
Wasn't aware that was a thing. I'll give it a shot.